pkg/controller/ads/dns.go

```diff
@@ -0,0 +1,317 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ads
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+	"slices"
+	"sync"
+	"time"
+
+	clusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	endpointv3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	core_v2 "kmesh.net/kmesh/api/v2/core"
+	"kmesh.net/kmesh/pkg/dns"
+)
+
+// adsDnsResolver is DNS resolver of Kernel Native
+type dnsController struct {
+	Clusters    chan []*clusterv3.Cluster
+	cache       *AdsCache
+	dnsResolver *dns.DNSResolver
+	// Store the copy of pendingResolveDomain.
+	clusterCache map[string]*pendingResolveDomain
+	// store all pending hostnames in the clusters
+	pendingClusterInfo map[string][]string
+	sync.RWMutex
+}
+
+// pending resolve domain info of Kennel-Native Mode,
+// domain name is used for dns resolution
+// cluster is used for create the apicluster
+type pendingResolveDomain struct {
+	DomainName  string
+	Clusters    []*clusterv3.Cluster
+	RefreshRate time.Duration
+}
+
+func NewDnsResolver(adsCache *AdsCache) (*dnsController, error) {
+	resolver, err := dns.NewDNSResolver()
+	if err != nil {
+		return nil, err
+	}
+	return &dnsController{
+		Clusters:           make(chan []*clusterv3.Cluster),
+		cache:              adsCache,
+		dnsResolver:        resolver,
+		clusterCache:       make(map[string]*pendingResolveDomain),
+		pendingClusterInfo: make(map[string][]string),
+	}, nil
+}
+
+func (r *dnsController) Run(stopCh <-chan struct{}) {
+	// Start dns resolver
+	go r.dnsResolver.StartDnsResolver(stopCh)
+	// Handle cds updates when a hostname completes resolution
+	go r.refreshAdsWorker(stopCh)
+	// Consumption of clusters to be resolved.
+	go r.startDnsController()
+	go func() {
+		<-stopCh
+		close(r.Clusters)
+	}()
+}
+
+func (r *dnsController) startDnsController() {
+	for clusters := range r.Clusters {
+		r.resolveDomains(clusters)
+	}
+}
+
+func (r *dnsController) resolveDomains(cds []*clusterv3.Cluster) {
+	domains, hostNames := getPendingResolveDomain(cds)
+
+	// store all pending hostnames of clusters in r.hostInfo
+	for _, cluster := range cds {
+		clusterName := cluster.GetName()
+		info := getHostName(cluster)
+		r.pendingClusterInfo[clusterName] = info
+	}
+
+	// delete any scheduled re-resolve for domains we no longer care about
+	r.dnsResolver.RemoveUnwatchDomain(hostNames)
+
+	// Update clusters based on the data in the dns cache.
+	for k, v := range domains {
+		addresses := r.dnsResolver.GetDNSAddresses(k)
+		// Already have record in dns cache
+		if addresses != nil {
+			r.updateClusters(v, addresses)
+			go r.cache.ClusterCache.Flush()
```

Comment:
Is this right? You are getting one domain resolved address and then flush the cluster